tests/blockchain/blockchain_test_utils.py

```diff
@@ -0,0 +1,142 @@
+                # make sure we don't have any duplicate heights. Each block
+                # height can only have a single block with in_main_chain set
+                assert height not in in_chain
+                in_chain.add(height)
+                if height > max_height:
+                    max_height = height
+
+        # make sure every height is represented in the set
+        assert len(in_chain) == max_height + 1
+
+
+async def _validate_and_add_block(
+    blockchain: Blockchain,
+    block: FullBlock,
+    expected_result: Optional[ReceiveBlockResult] = None,
+    expected_error: Optional[Err] = None,
+    skip_prevalidation: bool = False,
+) -> None:
+    # Tries to validate and add the block, and checks that there are no errors in the process and that the
+    # block is added to the peak.
+    # If expected_result is not None, that result will be enforced.
+    # If expected_error is not None, that error will be enforced (unless prevalidation fails, in which
+    # case we return with no errors). If expected_error is not None, receive_block must return Err.INVALID_BLOCK.
+    # If expected_result == INVALID_BLOCK but expected_error is None, we will allow for errors to happen
+
+    await check_block_store_invariant(blockchain)
+    if skip_prevalidation:
+        results = PreValidationResult(None, uint64(1), None, False)
+    else:
+        pre_validation_results: Optional[
+            List[PreValidationResult]
+        ] = await blockchain.pre_validate_blocks_multiprocessing([block], {}, validate_signatures=False)
+
+        if pre_validation_results is None:
+            # Returning None from pre validation means an error occurred
+            if expected_error is not None or expected_result == ReceiveBlockResult.INVALID_BLOCK:
+                return None
+            else:
+                raise ValueError("Prevalidation returned None")
+        results = pre_validation_results[0]
+    if results.error is not None:
+        if expected_result == ReceiveBlockResult.INVALID_BLOCK:
+            await check_block_store_invariant(blockchain)
+            return None
+        if expected_error is None:
+            raise ValueError(Err(results.error))
+        elif Err(results.error) != expected_error:
+            raise ValueError(f"Expected {expected_error} but got {Err(results.error)}")
+        await check_block_store_invariant(blockchain)
+        return None
+
+    result, err, _, _ = await blockchain.receive_block(block, results)
+    await check_block_store_invariant(blockchain)
+    if expected_error is None and expected_result != ReceiveBlockResult.INVALID_BLOCK:
+        # Not expecting any errors
+        if err is not None:
+            # Got an error
+            raise ValueError(err)
+    else:
+        # Expecting an error
+        if err != expected_error:
+            # Did not get the right error, or did not get an error
+            raise ValueError(f"Expected {expected_error} but got {err}")
```

Comment:
is there a reason to not use `assert` here, rather than throwing `ValueError`?
in my mind at least, an `AssertionError` is a hard failure and means the test failed, whereas a `ValueError` might be expected for negative tests. I don't think these errors can ever mean the test still passes, right?

Reply:
yeah you are right, makes sense to change to assertion error